nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -181,6 +218,244 @@ impl DataStore {
             .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
     }
 
+    async fn volume_checkout_allowed(
+        reason: &VolumeCheckoutReason,
+        vcr: &VolumeConstructionRequest,
+        maybe_disk: Option<Disk>,
+        maybe_instance: Option<Instance>,
+    ) -> Result<(), VolumeGetError> {
+        match reason {
+            VolumeCheckoutReason::ReadOnlyCopy => {
+                // When checking out to make a copy (usually for use as a
+                // read-only parent), the volume must be read only. Even if a
+                // call-site that uses Copy sends this copied Volume to a
+                // Propolis or Pantry, the Upstairs that will be created will be
+                // read-only, and will not take over from other read-only
+                // Upstairs.
+
+                match volume_is_read_only(&vcr) {
+                    Ok(read_only) => {
+                        if !read_only {
+                            return Err(VolumeGetError::CheckoutConditionFailed(
+                                String::from("Non-read-only Volume Checkout for use Copy!")
+                            ));
+                        }
+
+                        Ok(())
+                    }
+
+                    Err(e) => Err(VolumeGetError::InvalidVolume(e.to_string())),
+                }
+            }
+
+            VolumeCheckoutReason::CopyAndModify => {
+                // `CopyAndModify` is used when taking a read/write Volume,
+                // modifying it (for example, when taking a snapshot, to point
+                // to read-only resources), and committing it back to the DB.
+                // This is a checkout of a read/write Volume, so creating an
+                // Upstairs from it *may* take over from something else. The
+                // call-site must ensure this doesn't happen, but we can't do
+                // that here.
+
+                Ok(())
+            }
+
+            VolumeCheckoutReason::InstanceStart { vmm_id } => {
+                // Check out this volume to send to Propolis to start an
+                // Instance. The VMM id in the enum must match the instance's
+                // propolis_id.
```

Comment:
Can a second volume checkout to a new/different `vmm_id` come through here?
I guess I'm asking because I want to know how does this volume_checkout work with instance lifecycle.
If we have a VM that has crashed that had checked out a volume, can Nexus handle that itself after we have picked up the pieces and deleted the old VM?  I want to be sure this does not introduce a new additional "lock" on a volume that we have to clean up when things get stuck.

Reply:
In the `VolumeCheckoutReason::InstanceStart { vmm_id }` code path, the instance object and runtime are queried - if the old VM is deleted then there shouldn't be any conflicts when the new instance is created. I imagine there'd be a problem in the instance create saga if so, long before the instance start saga's volume checkout.